// @moonwell-fi/moonwell-sdk/_cjs/environments/abis/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MaximillionAbi = exports.WrappedNativeTokenAbi = exports.VoteCollectorAbi = exports.TokenSaleAbi = exports.TokenAbi = exports.TemporalGovernorAbi = exports.StakingTokenAbi = exports.MultichainGovernorAbi = exports.MultiRewardDistributorAbi = exports.MorphoViewsAbi = exports.MorphoVaultAbi = exports.MorphoPublicAllocatorAbi = exports.MorphoBundlerAbi = exports.MorphoBlueAbi = exports.MarketTokenAbi = exports.GovernorAbi = exports.GovernanceTokenAbi = exports.CoreViewsAbi = exports.CoreRouterAbi = exports.ComptrollerTokenAbi = exports.ChainLinkOracleAbi = void 0;
const chainLinkOracleAbi_js_1 = __importDefault(require("./chainLinkOracleAbi.js"));
exports.ChainLinkOracleAbi = chainLinkOracleAbi_js_1.default;
const comptrollerAbi_js_1 = __importDefault(require("./comptrollerAbi.js"));
exports.ComptrollerTokenAbi = comptrollerAbi_js_1.default;
const coreRouterAbi_js_1 = __importDefault(require("./coreRouterAbi.js"));
exports.CoreRouterAbi = coreRouterAbi_js_1.default;
const coreViewsAbi_js_1 = __importDefault(require("./coreViewsAbi.js"));
exports.CoreViewsAbi = coreViewsAbi_js_1.default;
const governanceTokenAbi_js_1 = __importDefault(require("./governanceTokenAbi.js"));
exports.GovernanceTokenAbi = governanceTokenAbi_js_1.default;
const governorAbi_js_1 = __importDefault(require("./governorAbi.js"));
exports.GovernorAbi = governorAbi_js_1.default;
const marketTokenAbi_js_1 = __importDefault(require("./marketTokenAbi.js"));
exports.MarketTokenAbi = marketTokenAbi_js_1.default;
const maximillionAbi_js_1 = __importDefault(require("./maximillionAbi.js"));
exports.MaximillionAbi = maximillionAbi_js_1.default;
const morphoBlueAbi_js_1 = __importDefault(require("./morphoBlueAbi.js"));
exports.MorphoBlueAbi = morphoBlueAbi_js_1.default;
const morphoBundlerAbi_js_1 = __importDefault(require("./morphoBundlerAbi.js"));
exports.MorphoBundlerAbi = morphoBundlerAbi_js_1.default;
const morphoPublicAllocatorAbi_js_1 = __importDefault(require("./morphoPublicAllocatorAbi.js"));
exports.MorphoPublicAllocatorAbi = morphoPublicAllocatorAbi_js_1.default;
const morphoVaultAbi_js_1 = __importDefault(require("./morphoVaultAbi.js"));
exports.MorphoVaultAbi = morphoVaultAbi_js_1.default;
const morphoViewsAbi_js_1 = __importDefault(require("./morphoViewsAbi.js"));
exports.MorphoViewsAbi = morphoViewsAbi_js_1.default;
const multiRewardDistributorAbi_js_1 = __importDefault(require("./multiRewardDistributorAbi.js"));
exports.MultiRewardDistributorAbi = multiRewardDistributorAbi_js_1.default;
const multichainGovernorAbi_js_1 = __importDefault(require("./multichainGovernorAbi.js"));
exports.MultichainGovernorAbi = multichainGovernorAbi_js_1.default;
const stakingTokenAbi_js_1 = __importDefault(require("./stakingTokenAbi.js"));
exports.StakingTokenAbi = stakingTokenAbi_js_1.default;
const temporalGovernorAbi_js_1 = __importDefault(require("./temporalGovernorAbi.js"));
exports.TemporalGovernorAbi = temporalGovernorAbi_js_1.default;
const tokenAbi_js_1 = __importDefault(require("./tokenAbi.js"));
exports.TokenAbi = tokenAbi_js_1.default;
const tokenSaleAbi_js_1 = __importDefault(require("./tokenSaleAbi.js"));
exports.TokenSaleAbi = tokenSaleAbi_js_1.default;
const voteCollectorAbi_js_1 = __importDefault(require("./voteCollectorAbi.js"));
exports.VoteCollectorAbi = voteCollectorAbi_js_1.default;
const wrappedNativeTokenAbi_js_1 = __importDefault(require("./wrappedNativeTokenAbi.js"));
exports.WrappedNativeTokenAbi = wrappedNativeTokenAbi_js_1.default;
