// @moonwell-fi/moonwell-sdk/_cjs/environments/definitions/ethereum/environment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.tokens = exports.ethereum = exports.createEnvironment = void 0;
const viem_1 = require("viem");
const chains_1 = require("viem/chains");
Object.defineProperty(exports, "ethereum", { enumerable: true, get: function () { return chains_1.mainnet; } });
const config_js_1 = require("../../types/config.js");
const tokens_js_1 = require("./tokens.js");
Object.defineProperty(exports, "tokens", { enumerable: true, get: function () { return tokens_js_1.tokens; } });
const createEnvironment = (rpcUrls, indexerUrl, governanceIndexerUrl) => (0, config_js_1.createEnvironmentConfig)({
    key: "ethereum",
    name: "Ethereum",
    chain: chains_1.mainnet,
    transport: rpcUrls
        ? (0, viem_1.fallback)(rpcUrls.map((url) => (0, viem_1.http)(url)))
        : (0, viem_1.http)(chains_1.mainnet.rpcUrls.default.http[0]),
    indexerUrl: indexerUrl || "https://ponder.moonwell.fi",
    governanceIndexerUrl: governanceIndexerUrl || "https://ponder.moonwell.fi",
    tokens: tokens_js_1.tokens,
    markets: {},
    vaults: {},
    morphoMarkets: {},
    contracts: {},
    custom: {},
});
exports.createEnvironment = createEnvironment;
