// @moonwell-fi/moonwell-sdk/_esm/actions/core/user-positions/getUserPositionSnapshots.js
import axios from "axios";
import dayjs from "dayjs";
import utc from "dayjs/plugin/utc.js";
import { getEnvironmentFromArgs, isStartOfDay } from "../../../common/index.js";
dayjs.extend(utc);
export async function getUserPositionSnapshots(client, args) {
    const environment = getEnvironmentFromArgs(client, args);
    if (!environment) {
        return [];
    }
    return fetchUserPositionSnapshots(args.userAddress, environment);
}
async function fetchUserPositionSnapshots(userAddress, environment) {
    const dailyData = [];
    let hasNextPage = true;
    let endCursor;
    while (hasNextPage) {
        const result = await axios.post(environment.indexerUrl, {
            query: `
          query {
            accountDailySnapshots(
              limit: 365,
              orderDirection: "desc",
              orderBy: "timestamp",
              where: { accountAddress: "${userAddress.toLowerCase()}", chainId: ${environment.chainId} }
              ${endCursor ? `after: "${endCursor}"` : ""}
            ) {
              items {
                timestamp,
                totalBorrowsUSD,
                totalSuppliesUSD,
                totalCollateralUSD,
              }
              pageInfo {
                hasNextPage
                endCursor
              }
            }
          }
        `,
        });
        dailyData.push(...result.data.data.accountDailySnapshots.items.filter((f) => isStartOfDay(f.timestamp)));
        hasNextPage = result.data.data.accountDailySnapshots.pageInfo.hasNextPage;
        endCursor = result.data.data.accountDailySnapshots.pageInfo.endCursor;
    }
    if (dailyData.length > 0) {
        return dailyData.map((point) => {
            const borrowUsd = Number(point.totalBorrowsUSD);
            const suppliedUsd = Number(point.totalSuppliesUSD);
            const collateralUsd = Number(point.totalCollateralUSD);
            const result = {
                chainId: environment.chainId,
                timestamp: point.timestamp * 1000,
                totalSupplyUsd: suppliedUsd,
                totalBorrowsUsd: borrowUsd,
                totalCollateralUsd: collateralUsd,
            };
            return result;
        });
    }
    else {
        return [];
    }
}
