// @moonwell-fi/moonwell-sdk/_cjs/actions/core/getUserBalances.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getUserBalances = void 0;
const viem_1 = require("viem");
const index_js_1 = require("../../common/index.js");
const index_js_2 = require("../../environments/utils/index.js");
const getTokenBalance = async (environment, userAddress, tokenAddress) => {
    if (tokenAddress === viem_1.zeroAddress) {
        return new Promise((resolve) => {
            environment.publicClient
                .getBalance({
                address: userAddress,
            })
                .then((balance) => {
                resolve({ amount: BigInt(balance), token: tokenAddress });
            })
                .catch(() => {
                resolve({ amount: 0n, token: tokenAddress });
            });
        });
    }
    const erc20Abi = (0, viem_1.parseAbi)([
        "function balanceOf(address owner) view returns (uint256)",
    ]);
    const erc20Contract = (0, viem_1.getContract)({
        address: tokenAddress,
        abi: erc20Abi,
        client: environment.publicClient,
    });
    const result = new Promise((resolve) => {
        erc20Contract.read
            .balanceOf([userAddress])
            .then((balance) => {
            resolve({ amount: BigInt(balance), token: tokenAddress });
        })
            .catch(() => {
            resolve({ amount: 0n, token: tokenAddress });
        });
    });
    return result;
};
async function getUserBalances(client, args) {
    const { userAddress } = args;
    const environments = (0, index_js_1.getEnvironmentsFromArgs)(client, args, false);
    const environmentsTokensBalances = await Promise.all(environments.map((environment) => {
        if (environment.contracts.views) {
            return Promise.all([
                environment.contracts.views.read.getTokensBalances([
                    Object.values(environment.config.tokens).map((token) => token.address),
                    userAddress,
                ]),
            ]);
        }
        return Promise.all([
            Promise.all(Object.values(environment.config.tokens).map((token) => getTokenBalance(environment, userAddress, token.address))),
        ]);
    }));
    const result = environments.flatMap((env, index) => {
        const balances = environmentsTokensBalances[index][0];
        const userBalances = balances
            .map((balance) => {
            const token = (0, index_js_2.findTokenByAddress)(env, balance.token);
            if (token) {
                const result = {
                    chainId: env.chainId,
                    account: userAddress,
                    token,
                    tokenBalance: new index_js_1.Amount(balance.amount, token.decimals),
                };
                return result;
            }
            else {
                return;
            }
        })
            .filter((item) => item !== undefined);
        return userBalances;
    });
    return result;
}
exports.getUserBalances = getUserBalances;
