// @moonwell-fi/moonwell-sdk/_cjs/environments/definitions/optimism/environment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.custom = exports.contracts = exports.morphoMarkets = exports.vaults = exports.tokens = exports.markets = exports.createEnvironment = void 0;
const viem_1 = require("viem");
const chains_1 = require("viem/chains");
const config_js_1 = require("../../types/config.js");
const contracts_js_1 = require("./contracts.js");
Object.defineProperty(exports, "contracts", { enumerable: true, get: function () { return contracts_js_1.contracts; } });
const core_markets_js_1 = require("./core-markets.js");
Object.defineProperty(exports, "markets", { enumerable: true, get: function () { return core_markets_js_1.markets; } });
const custom_js_1 = require("./custom.js");
Object.defineProperty(exports, "custom", { enumerable: true, get: function () { return custom_js_1.custom; } });
const morpho_markets_js_1 = require("./morpho-markets.js");
Object.defineProperty(exports, "morphoMarkets", { enumerable: true, get: function () { return morpho_markets_js_1.morphoMarkets; } });
const morpho_vaults_js_1 = require("./morpho-vaults.js");
Object.defineProperty(exports, "vaults", { enumerable: true, get: function () { return morpho_vaults_js_1.vaults; } });
const tokens_js_1 = require("./tokens.js");
Object.defineProperty(exports, "tokens", { enumerable: true, get: function () { return tokens_js_1.tokens; } });
const optimism = (0, viem_1.defineChain)({ ...chains_1.optimism, testnet: false });
const createEnvironment = (rpcUrls, indexerUrl, governanceIndexerUrl) => (0, config_js_1.createEnvironmentConfig)({
    key: "optimism",
    name: "OP Mainnet",
    chain: optimism,
    transport: rpcUrls
        ? (0, viem_1.fallback)(rpcUrls.map((url) => (0, viem_1.http)(url)))
        : (0, viem_1.http)(optimism.rpcUrls.default.http[0]),
    indexerUrl: indexerUrl || "https://ponder.moonwell.fi",
    governanceIndexerUrl: governanceIndexerUrl || "https://ponder.moonwell.fi",
    tokens: tokens_js_1.tokens,
    markets: core_markets_js_1.markets,
    vaults: morpho_vaults_js_1.vaults,
    morphoMarkets: morpho_markets_js_1.morphoMarkets,
    contracts: contracts_js_1.contracts,
    custom: custom_js_1.custom,
});
exports.createEnvironment = createEnvironment;
